### PR TITLE
rustdoc: make `--passes` and `--no-defaults` have no effect

### LINKED ISSUE #91714: Make `rustdoc --passes` and `rustdoc --no-defaults` have no effect
Currently, `rustdoc --passes` does one of two things:
1. `--passes list` shows a list of available passes:
```
Available passes for running rustdoc:
check_doc_test_visibility - run various visibility-related lints on doctests
        strip-hidden - strips all `#[doc(hidden)]` items from the output
   unindent-comments - removes excess indentation on comments in order for markdown to like it
       strip-private - strips all private items from a crate which cannot be seen externally, implies strip-priv-imports
  strip-priv-imports - strips all private import statements (`use`, `extern crate`) from a crate
   propagate-doc-cfg - propagates `#[doc(cfg(...))]` to child items
collect-intra-doc-links - resolves intra-doc links
check-code-block-syntax - validates syntax inside Rust code blocks
 collect-trait-impls - retrieves trait impls for items in the crate
calculate-doc-coverage - counts the number of items with and without documentation
check-invalid-html-tags - detects invalid HTML tags in doc comments
     check-bare-urls - detects URLs that are not hyperlinks

Default passes for rustdoc:
 collect-trait-impls
   unindent-comments
check_doc_test_visibility
        strip-hidden  (when not --document-hidden-items)
       strip-private  (when not --document-private-items)
  strip-priv-imports  (when --document-private-items)
collect-intra-doc-links
check-code-block-syntax
check-invalid-html-tags
   propagate-doc-cfg
     check-bare-urls

Passes run with `--show-coverage`:
        strip-hidden  (when not --document-hidden-items)
       strip-private  (when not --document-private-items)
calculate-doc-coverage
```
2. `rustdoc --passes collect-trait-impls` will add that specific pass to the list of passes rustdoc executes (I think this is intended to be used with `--no-defaults`, which is also deprecated).

Both of these are stable, so unfortunately we can't remove them; 2. is deprecated and 1. should be soon (after https://github.com/rust-lang/rust/issues/91713 is fixed). However, this is not a very useful sort of stability, because the names and number of passes is not stable, and rustdoc is very likely to be buggy if you don't run all passes (there's no sort of test for this).

I propose not just deprecating `--passes` and `--no-defaults` but making them a no-op altogether, which means less complexity both for users and in the rustdoc codebase. I also suggest removing the CLI help at the same time (and just say "this flag is a no-op").

Note that the only other deprecated flag, `--input-format`, is also currently a no-op when used with `--input-format rust`, and a hard error otherwise. We could go further and make everything but `--passes=list` be a hard error, but that seems like more of a breaking change than necessary for the goal of reducing complexity.

### PR DESCRIPTION
**Pull request by pitaj**
Fixes #91714

One potential issue is that currently there is no stable way to achieve `--document-hidden-items`. This affects test `issue-15347`.

I also had to modify the tests `issue-42875` and `no-compiler-export`. Regardless of combinations of `--document-hidden-items` and `--document-private-items`, I was unable to get these to pass without the modifications. I left behind a comment noting the change.

### CODE REVIEW

File: src/test/rustdoc/no-compiler-reexport.rs
Comment (jyn514) [inline src/test/rustdoc/no-compiler-reexport.rs:4]: This seems fine to me, it makes the behavior match that the injected re-exports are hidden too. I want to take a look at the git history of the test before merging though.

**Review by jyn514 — COMMENTED** [review]
This looks great, thanks!

File: src/test/rustdoc/no-compiler-reexport.rs
Comment (jyn514) [reply src/test/rustdoc/no-compiler-reexport.rs:4]: Can you add a `!has` check here to make sure the extern crate statement doesn't appear in the docs?

Comment (jyn514) [reply src/test/rustdoc/no-compiler-reexport.rs:4]: Looks like it was added in https://github.com/rust-lang/rust/pull/77785 - @GuillaumeGomez is there any particular reason you wanted `extern crate std;` to appear when documenting with `--document-hidden-items`? I don't think it's related to that flag in any meaningful way.

Comment (GuillaumeGomez) [reply src/test/rustdoc/no-compiler-reexport.rs:4]: This is surprising indeed. No I don't remember at all why I would have wanted to have the `extern crate std;` displayed...

File: src/test/rustdoc/issue-42875.rs
Comment (pitaj) [reply src/test/rustdoc/issue-42875.rs]: Removed.

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

File: src/test/rustdoc/no-compiler-reexport.rs
Comment (jyn514) [inline src/test/rustdoc/no-compiler-reexport.rs:6]: I would just remove this comment here, I don't think it's necessary. Thanks for looking into it though!

File: src/librustdoc/config.rs
Comment (jyn514) [inline src/librustdoc/config.rs:780]: I think the backquotes were fine, no need to change them.
```suggestion
            let mut err = diag.struct_warn(&format!("the `{}` flag no longer functions", flag));
```

File: src/librustdoc/core.rs
Comment (jyn514) [inline src/librustdoc/core.rs:428]: Ugh, not super happy that this can be configured both through attributes and a CLI flags ... but that's a fight for another day.

File: src/librustdoc/core.rs
Comment (jyn514) [inline src/librustdoc/core.rs:444]: nit: you can combine this into one condition:
```suggestion
        if attr.is_word() && name == sym::no_default_passes {
```

File: src/librustdoc/config.rs
Comment (pitaj) [reply src/librustdoc/config.rs:780]: I didn't change them, it matches the format used for removed flags before, but I can certainly align removed flags and deprecated flags to both use backticks.

File: src/librustdoc/passes/mod.rs
Comment (camelid) [inline src/librustdoc/passes/mod.rs:134]: In a later PR, I feel like we should get rid of this enum altogether. It doesn't need to happen though.

**Review by jyn514 — APPROVED** [review]
This looks great! Can you just squash the commits into one? Then it looks to good to merge once the FCP in https://github.com/rust-lang/rust/issues/91714 finishes :)

**Comment by camelid** [comment]
FCP has completed! @rustbot ready

**Comment by jyn514** [comment]
Thanks for working on this!

**Pull request merged**
